### PR TITLE
Keep origin of contract call stack

### LINKED ISSUE #34: Keep extrinsic sender through contract call stack
Provide the means to query the extrinsic sender address during nested contract execution.
So that the extrinsic origin is known throughout nested contract calls.

For reference, `caller()` gives the address of the calling account (extrinsic sender at the top level, changes to calling contract for nested calls)
https://github.com/plugblockchain/plug-blockchain/blob/b93e3585eb9bb9f1648a0d155e67832a8aa7f30c/srml/contracts/src/exec.rs#L139-L140

unblocks: #32

### PR DESCRIPTION
**Pull request by aliXsed**
Closes #34

### CODE REVIEW

**Comment by aliXsed** [comment]
@holygits This is my understanding of the task. Please correct me if I'm wrong.

**Comment by aliXsed** [comment]
@holygits Also do we need to add some unit tests to check this?

**Comment by jordy25519** [comment]
Yeah this is it. There should be tests for
- origin is correct at top level call
- origin is correct for nested call

File: frame/contracts/src/exec.rs [hunk 1]
```diff
@@ -293,10 +294,12 @@ where
 	/// The specified `origin` address will be used as `sender` for. The `origin` must be a regular
 	/// account (not a contract).
 	pub fn top_level(origin: T::AccountId, cfg: &'a Config<T>, vm: &'a V, loader: &'a L) -> Self {
+		let source_account = origin;
+		let original_account = source_account.clone();
 		ExecutionContext {
 			parent: None,
 			self_trie_id: None,
-			self_account: origin,
+			self_account: source_account,
```
Comment (jordy25519) [inline frame/contracts/src/exec.rs]: I think `origin.clone()` would be clearer

File: frame/contracts/src/exec.rs [hunk 2]
```diff
@@ -674,6 +681,7 @@ struct CallContext<'a, 'b: 'a, T: Trait + 'b, V: Vm<T> + 'b, L: Loader<T>> {
 	value_transferred: BalanceOf<T>,
 	timestamp: MomentOf<T>,
 	block_number: T::BlockNumber,
+	origin: T::AccountId, // origin has the same value as caller except for nested calls
```
Comment (jordy25519) [inline frame/contracts/src/exec.rs:685]: Can be a doc comment: ```///  `origin` is the account which authored the original extrinsic leading to the current contract call(s). It does not change on nested calls as with `caller` ```

File: frame/contracts/src/exec.rs (hunk 1 above)
Comment (aliXsed) [reply frame/contracts/src/exec.rs]: I wanted to call only one clone, but I agree origin.clone would be clearer.

**Comment by cowboy-bebug** [comment]
For the build issue, we need to resolve this error:
```
   Compiling pallet-elections-phragmen v2.0.0 (/Users/ericlim/project/plug-blockchain-2/frame/elections-phragmen)
   Compiling sp-consensus-babe v2.0.0 (/Users/ericlim/project/plug-blockchain-2/primitives/consensus/babe)
   Compiling pallet-contracts v2.0.0 (/Users/ericlim/project/plug-blockchain-2/frame/contracts)
error[E0407]: method `origin` is not a member of trait `Ext`
   --> /Users/ericlim/project/plug-blockchain-2/frame/contracts/src/exec.rs:763:2
    |
763 |       fn origin(&self) -> &T::AccountId {
    |  _____^
764 | |         &self.origin
765 | |     }
    | |_____^ not a member of trait `Ext`
```

**Review by cowboy-bebug — CHANGES_REQUESTED** [review]
Nice and clean! Just need to add `fn origin()` to the `Ext` trait to resolve build issue:
https://github.com/plugblockchain/plug-blockchain/blob/0966e9d414e018f7a1ab85b6ed145eb78af590a6/frame/contracts/src/exec.rs#L96

**Comment by aliXsed** [comment]
> Nice and clean! Just need to add `fn origin()` to the `Ext` trait to resolve build issue:
> https://github.com/plugblockchain/plug-blockchain/blob/0966e9d414e018f7a1ab85b6ed145eb78af590a6/frame/contracts/src/exec.rs#L96

Thank you Eric! I will do. I created the PR early even before building it as a draft PR to make sure I'm on the right track.

File: frame/contracts/src/exec.rs (hunk 2 above)
Comment (aliXsed) [reply frame/contracts/src/exec.rs:685]: I added the doc to the public interface instead.

**Review by cowboy-bebug — CHANGES_REQUESTED** [review]
Looks good! I think we might want to add one more test for origin() like the tests below:
https://github.com/plugblockchain/plug-blockchain/blob/0966e9d414e018f7a1ab85b6ed145eb78af590a6/frame/contracts/src/wasm/mod.rs#L767
https://github.com/plugblockchain/plug-blockchain/blob/0966e9d414e018f7a1ab85b6ed145eb78af590a6/frame/contracts/src/wasm/mod.rs#L829

**Review by cowboy-bebug — APPROVED** [review]
Yep that's it! Nice work!! 👍

**Pull request merged**